aten/src/ATen/native/cuda/ReduceOpsKernel.cu

```diff
@@ -222,6 +224,227 @@ void argmin_kernel_cuda(TensorIterator& iter) {
+ * considered as having 'num_rows' rows of size 'row_size'.
+ * Each thread block processes one or more sets of contiguous rows (processing multiple rows
+ * per thread block is quicker than processing a single row, especially for short rows).
+ */
+template<typename scalar_t, int num_threads_x, int num_threads_y, class BinaryFunction>
+__global__ void tensor_kernel_scan_innermost_dim_with_indices(const scalar_t *self_, scalar_t *values_, int64_t *indices_,
+                                                int num_rows, int row_size,
+                                                scalar_t init, BinaryFunction binary_op) {
+  __shared__ scalar_t vbuf[num_threads_y][2 * num_threads_x];
+  __shared__ int64_t ibuf[num_threads_y][2 * num_threads_x];
+  scalar_t* row_buf = vbuf[threadIdx.y];
+  int64_t* row_idx_buf = ibuf[threadIdx.y];
+
+  for (int block_row = blockIdx.x * blockDim.y;
+       block_row < num_rows;
+       block_row += blockDim.y * gridDim.x) {
+    int row = block_row + threadIdx.y;
+    const scalar_t *row_self = self_ + row * row_size;
+    scalar_t *row_values = values_ + row * row_size;
+    int64_t *row_indices = indices_ + row * row_size;
+    scalar_t block_total = init;
+    int64_t block_idx_final = 0;
+    // Perform scan on one block at a time, keeping track of the total value of
+    // all blocks processed so far.
+    for (int block_col = 0; block_col < row_size; block_col += 2 * num_threads_x) {
+      // Load data into shared memory (two values per thread).
+      int col1 = block_col + threadIdx.x;
+      int col2 = block_col + num_threads_x + threadIdx.x;
+      if (row < num_rows) {
+        if (col1 < row_size) {
+          row_buf[threadIdx.x] = row_self[col1];
+          row_idx_buf[threadIdx.x] = col1;
+        } else {
+          row_buf[threadIdx.x] = init;
+          // No need to set the index here as the value in init will never be selected
+        }
+
+        if (col2 < row_size) {
+          row_buf[num_threads_x + threadIdx.x] = row_self[col2];
+          row_idx_buf[num_threads_x + threadIdx.x] = col2;
+        } else {
+          row_buf[num_threads_x + threadIdx.x] = init;
+          // No need to set the index here as the value in init will never be selected
+        }
+
+        // Add the total value of all previous blocks to the first value of this block.
+        if (threadIdx.x == 0) {
+          auto out = binary_op(row_buf[0], block_total);
+          if(out == block_total) {
+            row_idx_buf[0] = block_idx_final;
+          }
+          row_buf[0] = out;
+        }
+      }
+      __syncthreads();
+
+      // Parallel reduction (up-sweep).
+      for (int s = num_threads_x, d = 1; s >= 1; s >>= 1, d <<= 1) {
+        if (row < num_rows && threadIdx.x < s) {
+          int offset = (2 * threadIdx.x + 1) * d - 1;
+          auto out = binary_op(row_buf[offset], row_buf[offset + d]);
+          if(out == row_buf[offset]) {
```

Comment:
Please make sure that this logic works if you get tensors filled with nans, because comparison will always be false in this case, so row_idx_buf values may not be correctly propagated